File: backend/app/engines/spyware_detector.py
import re
import uuid
from typing import List
from app.data.threat_intelligence import KNOWN_SPYWARE_DOMAINS, KNOWN_SPYWARE_PROCESSES, KNOWN_MALWARE_HASHES
from app.models.schemas import ThreatFinding, DeviceBehaviorMetrics

HEURISTIC_PATTERNS = [
    {
        "regex": re.compile(r'/var/containers/Bundle/Application/[a-f0-9-]+/(?!com\.apple)', re.IGNORECASE),
        "name": "Sideloaded Suspicious Application Bundle",
        "severity": "medium",
        "desc": "Unsigned or sideloaded app bundle path detected in iOS application sandbox.",
        "remediation": "Audit enterprise profile signatures and remove unauthorized MDM configuration profiles."
    },
    {
        "regex": re.compile(r'shutdown\.log.*(?:libtouchregd|bh|roleaborede|com\.apple\.WebKit\.WebContent\.helper)', re.IGNORECASE),
        "name": "Pegasus Artifacts in iOS Shutdown Log",
        "severity": "critical",
        "desc": "Shutdown log exhibits sticky process anomalies consistent with NSO Group Pegasus zero-click persistence.",
        "remediation": "Immediate device isolation required. Back up sysdiagnose logs for forensic verification (MVT analysis)."
    },
    {
        "regex": re.compile(r'datausage\.sqlite.*(?:unknown|untitled|[a-z0-9]{16,}).*network', re.IGNORECASE),
        "name": "Stealth Background Data Exfiltration",
        "severity": "high",
        "desc": "Obfuscated process name recorded transmitting packets in cellular/Wi-Fi data usage database.",
        "remediation": "Capture network pcap and inspect routing table."
    },
    {
        "regex": re.compile(r'(?:microphone|camera).*(?:background|inactive|locked|headless)', re.IGNORECASE),
        "name": "Headless Sensor Eavesdropping",
        "severity": "critical",
        "desc": "Hardware microphone or camera driver activated without foreground user interaction or screen activation.",
        "remediation": "Power off device immediately. Suspect active zero-click surveillance implant."
    },
    {
        "regex": re.compile(r'(?:root|ca[-_]certificate).*install.*(?:untrusted|self-signed|profile)', re.IGNORECASE),
        "name": "Rogue CA Certificate Injection",
        "severity": "high",
        "desc": "Untrusted root authority injected into system trust store enabling full HTTPS MITM decryption.",
        "remediation": "Remove untrusted certificates from Settings -> General -> About -> Certificate Trust Settings."
    },
    {
        "regex": re.compile(r'(?:cydia|substrate|sileo|checkra1n|unc0ver|taurine|palera1n)', re.IGNORECASE),
        "name": "Jailbreak / Root Exploitation Found",
        "severity": "high",
        "desc": "Jailbreak framework signatures found, reducing core kernel protections (KPP/KTRR bypassed).",
        "remediation": "Restore device firmware using official signed DFU recovery image."
    }
]

def scan_text_for_iocs(content: str) -> List[ThreatFinding]:
    findings: List[ThreatFinding] = []
    text_lower = content.lower()

    # 1. Known malicious domain check
    for domain_entry in KNOWN_SPYWARE_DOMAINS:
        if domain_entry["pattern"].lower() in text_lower:
            findings.append(ThreatFinding(
                id=str(uuid.uuid4()),
                type="domain_ioc",
                match=domain_entry["pattern"],
                name=f"Spyware C2 Domain: {domain_entry['pattern']}",
                category=domain_entry["type"],
                severity=domain_entry["severity"],
                description=f"Matched known Pegasus/Predator C2 endpoint '{domain_entry['pattern']}'.",
                recommendation="Block domain at DNS/firewall tier. Quarantine associated endpoints immediately."
            ))

    # 2. Known process check
    for proc in KNOWN_SPYWARE_PROCESSES:
        if proc["name"].lower() in text_lower:
            findings.append(ThreatFinding(
                id=str(uuid.uuid4()),
                type="process_ioc",
                match=proc["name"],
                name=f"Disguised Spyware Process: {proc['name']}",
                category="Active Spyware Process",
                severity=proc["severity"],
                description=f"{proc['desc']}. Process name matches documented Pegasus/Predator payload.",
                recommendation="Terminate process, preserve crashdump memory images, and conduct offline forensic analysis."
            ))

    # 3. Hash matching
    for file_hash in KNOWN_MALWARE_HASHES:
        if file_hash.lower() in text_lower:
            findings.append(ThreatFinding(
                id=str(uuid.uuid4()),
                type="hash_match",
                match=f"{file_hash[:16]}...",
                name="Known Spyware Binary Hash Match",
                category="Malware Hash",
                severity="critical",
                description=f"File SHA-256 hash {file_hash[:32]}... matches known NSO Pegasus payload binary.",
                recommendation="Quarantine the target file and submit to incident response team."
            ))

    # 4. Heuristic pattern checks
    for hp in HEURISTIC_PATTERNS:
        match = hp["regex"].search(content)
        if match:
            findings.append(ThreatFinding(
                id=str(uuid.uuid4()),
                type="heuristic_anomaly",
                match=hp["name"],
                name=hp["name"],
                category="Heuristic Forensic Analysis",
                severity=hp["severity"],
                description=hp["desc"],
                recommendation=hp["remediation"],
                raw=match.group(0)[:120]
            ))

    return findings

def evaluate_device_metrics(metrics: DeviceBehaviorMetrics) -> List[ThreatFinding]:
    findings: List[ThreatFinding] = []

    if metrics.batteryDrain > 30:
        sev = "critical" if metrics.batteryDrain > 55 else "high"
        findings.append(ThreatFinding(
            id=str(uuid.uuid4()),
            type="behavior_metric",
            match=f"{metrics.batteryDrain}% / day battery drain",
            name="Abnormal Power Consumption Anomaly",
            category="Device Telemetry",
            severity=sev,
            description=f"Battery discharge rate of {metrics.batteryDrain}%/day exceeds baseline by >250%. Indicates unthrottled audio encoding or continuous background sensor capture.",
            recommendation="Inspect background execution time per app in system analytics."
        ))

    if metrics.dataUsage > 500:
        sev = "critical" if metrics.dataUsage > 1000 else "high"
        findings.append(ThreatFinding(
            id=str(uuid.uuid4()),
            type="behavior_metric",
            match=f"{metrics.dataUsage} MB / day network transfer",
            name="Excessive Outbound Exfiltration Flow",
            category="Data Exfiltration",
            severity=sev,
            description=f"Monitored device transmitted {metrics.dataUsage} MB over cellular/Wi-Fi during idle state.",
            recommendation="Inspect per-socket transmission stats in sysdiagnose netstat tables."
        ))

    if metrics.cpuSpikes > 5:
        findings.append(ThreatFinding(
            id=str(uuid.uuid4()),
            type="behavior_metric",
            match=f"{metrics.cpuSpikes} unexplained CPU spikes",
            name="Unexplained CPU Burst Activity",
            category="Processor Anomalies",
            severity="medium",
            description=f"Identified {metrics.cpuSpikes} recurring high-load CPU spikes while the device display was locked.",
            recommendation="Analyze spin logs and core dumps to identify executing thread symbols."
        ))

    if metrics.unexpectedReboots > 2:
        findings.append(ThreatFinding(
            id=str(uuid.uuid4()),
            type="behavior_metric",
            match=f"{metrics.unexpectedReboots} panics / reboots",
            name="Kernel Panic / Exploit Triggered Crashes",
            category="Kernel Stability",
            severity="high",
            description=f"{metrics.unexpectedReboots} unexpected reboots or kernel panics detected. Often caused by failed heap spray or race condition exploits.",
            recommendation="Extract panic.ips files from DiagnosticReports and analyze register state."
        ))

    return findings

File: frontend/src/engines/spywareDetector.js
// Spyware & IOC Detection Engine
import { KNOWN_IOCS } from '../utils/threatData';

export function scanForIOCs(input) {
  const results = [];
  const text = input.toLowerCase();

  // Scan for known malicious domains
  KNOWN_IOCS.domains.forEach((ioc) => {
    if (text.includes(ioc.pattern.toLowerCase())) {
      results.push({
        type: 'domain',
        match: ioc.pattern,
        category: ioc.type,
        severity: ioc.severity,
        description: `Known ${ioc.type} domain detected: ${ioc.pattern}`,
        recommendation: 'Immediately isolate the device and perform a full forensic analysis.',
      });
    }
  });

  // Scan for known malicious process names
  KNOWN_IOCS.processes.forEach((proc) => {
    if (text.includes(proc.name.toLowerCase())) {
      results.push({
        type: 'process',
        match: proc.name,
        category: proc.desc,
        severity: proc.severity,
        description: `Malicious process "${proc.name}" detected — ${proc.desc}`,
        recommendation: 'This process is associated with advanced spyware. Terminate immediately and preserve logs for forensic investigation.',
      });
    }
  });

  // Scan for known file hashes
  KNOWN_IOCS.fileHashes.forEach((hash) => {
    if (text.includes(hash.toLowerCase())) {
      results.push({
        type: 'hash',
        match: hash.substring(0, 16) + '...',
        category: 'Known Malware Hash',
        severity: 'critical',
        description: `File hash matches known spyware binary: ${hash.substring(0, 32)}...`,
        recommendation: 'The matched file is confirmed malware. Quarantine immediately.',
      });
    }
  });

  // Heuristic checks
  const heuristicPatterns = [
    { regex: /\/var\/containers\/Bundle\/Application\/[a-f0-9-]+\/(?!com\.apple)/i, name: 'Suspicious App Bundle Path', severity: 'medium', desc: 'Unusual application bundle path detected — may indicate sideloaded malware' },
    { regex: /shutdown\.log.*(?:libtouchregd|bh|roleaborede)/i, name: 'Pegasus in Shutdown Log', severity: 'critical', desc: 'Pegasus indicator found in iOS shutdown log — device likely compromised' },
    { regex: /datausage\.sqlite.*(?:unknown|untitled).*network/i, name: 'Hidden Network Process', severity: 'high', desc: 'Unknown process with network activity detected in usage database' },
    { regex: /(?:microphone|camera).*(?:background|inactive|locked)/i, name: 'Background Sensor Access', severity: 'critical', desc: 'Camera or microphone accessed while device is locked — surveillance indicator' },
    { regex: /root.*certificate.*install/i, name: 'Root Certificate Installation', severity: 'high', desc: 'Root certificate installed — potential man-in-the-middle attack setup' },
    { regex: /(?:jailbreak|cydia|substrate|sileo)/i, name: 'Jailbreak Detected', severity: 'high', desc: 'Jailbreak indicators found — device security compromised' },
  ];

  heuristicPatterns.forEach((hp) => {
    if (hp.regex.test(text)) {
      results.push({
        type: 'heuristic',
        match: hp.name,
        category: 'Behavioral Analysis',
        severity: hp.severity,
        description: hp.desc,
        recommendation: 'Investigate further. This pattern may indicate active surveillance.',
      });
    }
  });

  return results;
}

// Analyze device behavior metrics for anomalies
export function analyzeDeviceBehavior(metrics) {
  const findings = [];

  if (metrics.batteryDrain > 30) {
    findings.push({
      type: 'behavior',
      match: 'High Battery Drain',
      severity: metrics.batteryDrain > 50 ? 'critical' : 'high',
      description: `Battery drain rate of ${metrics.batteryDrain}%/day is ${metrics.batteryDrain > 50 ? 'critically' : 'abnormally'} high — possible background spyware activity`,
      recommendation: 'Check background processes and recently installed apps.',
    });
  }

  if (metrics.dataUsage > 500) {
    findings.push({
      type: 'behavior',
      match: 'Excessive Data Usage',
      severity: metrics.dataUsage > 1000 ? 'critical' : 'high',
      description: `Data usage of ${metrics.dataUsage}MB/day is abnormal — possible data exfiltration`,
      recommendation: 'Review network connections and data transfer logs.',
    });
  }

  if (metrics.cpuSpikes > 5) {
    findings.push({
      type: 'behavior',
      match: 'CPU Spike Anomaly',
      severity: 'medium',
      description: `${metrics.cpuSpikes} unexplained CPU spikes detected — may indicate encryption operations by spyware`,
      recommendation: 'Monitor which processes consume CPU during spikes.',
    });
  }

  if (metrics.unexpectedReboots > 2) {
    findings.push({
      type: 'behavior',
      match: 'Unexpected Reboots',
      severity: 'high',
      description: `${metrics.unexpectedReboots} unexpected device reboots — possible exploit-triggered crashes`,
      recommendation: 'Preserve crash logs and analyze for exploit signatures.',
    });
  }

  return findings;
}
